fetch.bs

```diff
@@ -2820,6 +2823,11 @@ given a <a for=/>request</a> <var>request</var>, run these steps:
  <li><p>Let <var>serializedOrigin</var> be the result of <a>byte-serializing a request origin</a>
  with <var>request</var>.
 
+ <li><p>If <var>request</var>'s <a for=request>current URL</a>'s <a for=url>origin</a>'s
+ <a for=origin>host</a> ends with "<code>.onion</code>" or "<code>.onion.</code>", and
+ is not <a>same origin</a> with <var>request</var>'s <a for=request>origin</a>, then set
+ <var>serializedOrigin</var> to `<code>null</code>`. [[ONION]]
```

Comment:
Due to "byte-serializing a request origin" we'll also end up with null if they are same origin but there's a cross-origin redirect in between. Ideally we'd test that, but I guess `.onion` will be hard to test.

Reply:
I've got these two no-cors manual test cases on my [test page](http://ixrdj3iwwhkuau5tby5jh3a536a2rdhpbdbu6ldhng43r47kim7a3lid.onion/referrer/onion.html) (you can also see it at http://fmarier.com/referrer/onion.html though it's meant to run from a `.onion`) :
- `example.onion` --307--> `example.onion` --307--> `example.com`
- `example.onion` --307--> `example.com` --307--> `example.onion`

For the first one, both Brave and Tor Browser send a `null` origin (and omit the referrer for that matter).

For the second one, however, Brave sends a `null` origin but the Tor browser sends the full Origin in that case.

I think a `null` Origin is the desired behavior here?

Reply:
I'm not sure I understand the examples, what's the starting origin? As long as you are same-origin I would expect a referrer to be included, but once you've gone cross-origin once, it should no longer be possible.

Reply:
The starting origin in all cases is `example.onion`. It issues two 307 redirects.

In the first example, it goes from `example.onion` to `example.onion/anotherpage` before ending on `example.com`.

In the second example, it goes from `example.onion` to `example.com` before ending on `example.onion/anotherpage`.

Reply:
So the user navigates to the starting origin? I guess then the scenarios would work. (Typically with fetching there's a starting document with some kind of authority and that would do the fetching, which could then result in a redirect. The initial 307 threw me off.)

Assuming the user does the navigation I'd expect:

1. example.onion 307 to example.onion (Origin could be included here, e.g., in case of a POST); example.onion 307 to example.com (it would be null here)
2. example.onion 307 to example.com (it would be null here); example.com to example.onion (it would continue to be null here, as it's been tainted by example.com)

Reply:
> So the user navigates to the starting origin?

Yes, that's correct. However, I just realized that I made a mistake in my description of the test cases, which probably threw you off.

What I meant was:

1. Navigate to `example.onion/index.html`.
2. Submit form (POST) to `example.onion/another.html`.
3. `example.onion/another.html` is a 307 redirect to `example.com`.
4. You end up on `example.com` with a `null` `Origin` header.

The second test case is:

1. Navigate to `example.onion/index.html`.
2. Submit form (POST) to `example.com`.
3. `example.com` is a 307 redirect to `example.onion/another.html`.
4. You end up on `example.onion/another.html` with a `null` `Origin` header.

I had the same expectations as you (regarding the `example.com` tainting due to the redirect) and that's what I implemented in Brave, but Tor Browser returns an `Origin` of `example.onion` in the second case.

Reply:
That seems like a pretty serious bug. @KershawChang @valenting ^^

Reply:
I've filed [bug 1742784](https://bugzilla.mozilla.org/show_bug.cgi?id=1742784) for this.